cli/crates/server/src/bun.rs

```diff
@@ -0,0 +1,174 @@
+use std::path::{Path, PathBuf};
+use std::process::Stdio;
+use std::sync::Arc;
+use thiserror::Error;
+use tokio::process::Command;
+use tokio::task::JoinError;
+
+#[derive(Error, Debug, Clone)]
+pub enum NpmCommandError {
+    #[error("working directory '{0}' not found")]
+    WorkingDirectoryNotFound(PathBuf),
+    #[error("working directory '{0}' cannot be read.\nCaused by: {1}")]
+    WorkingDirectoryCannotBeRead(PathBuf, Arc<std::io::Error>),
+    /// returned if npm cannot be found
+    #[error("could not find npm: {0}")]
+    NotFound(which::Error),
+
+    /// returned if any of the npm commands exits unsuccessfully
+    #[error("npm encountered an error: {0}")]
+    CommandError(Arc<std::io::Error>),
+
+    /// returned if any of the npm commands exits unsuccessfully
+    #[error("npm failed with output:\n{0}")]
+    OutputError(String),
+    /// returned if any of the bun commands exits unsuccessfully
+    #[error("bun encountered an error: {0}")]
+    BunCommandError(String),
+
+    /// returned if any of the bun commands exits unsuccessfully
+    #[error("bun failed with output:\n{0}")]
+    BunOutputError(String),
+}
+// this has to be clone for use in the config manager
+#[derive(Error, Debug, Clone)]
+pub enum BunError {
+    #[error("Could not create a lock for the bun installation: {0}")]
+    Lock(Arc<fslock::Error>),
+
+    #[error("Could not release the lock for the bun installation: {0}")]
+    Unlock(Arc<fslock::Error>),
+
+    /// returned if the directory cannot be read
+    #[error("could not create path '{0}' for the bun installation")]
+    CreateDir(PathBuf),
+
+    /// returned if a spawned task panics
+    #[error("{0}")]
+    SpawnedTaskPanic(Arc<JoinError>),
+
+    /// returned if any of the package manager commands ran during resolver build exits unsuccessfully
+    #[error("command error: {0}")]
+    BunInstallPackageManagerCommandError(#[from] NpmCommandError),
+}
+
+impl From<JoinError> for BunError {
+    fn from(value: JoinError) -> Self {
+        Self::SpawnedTaskPanic(Arc::new(value))
+    }
+}
+
+const BUN_VERSION: &str = "1.0.28";
+
+async fn run_command<P: AsRef<Path>>(
+    arguments: &[&str],
+    current_directory: P,
+) -> Result<Option<Vec<u8>>, NpmCommandError> {
+    let command_string = format!("npm {}", arguments.iter().format(" "));
+    let current_directory = current_directory.as_ref();
+    match current_directory.try_exists() {
+        Ok(true) => Ok(()),
+        Ok(false) => Err(NpmCommandError::WorkingDirectoryNotFound(current_directory.to_owned())),
+        Err(err) => Err(NpmCommandError::WorkingDirectoryCannotBeRead(
+            current_directory.to_owned(),
+            Arc::new(err),
+        )),
+    }?;
+    trace!("running '{command_string}'");
+
+    // Use `which` to work-around weird path search issues on Windows.
+    // See https://github.com/rust-lang/rust/issues/37519.
+    let program_path = which::which("npm").map_err(NpmCommandError::NotFound)?;
```

Comment:
Does this play nice if a user is using yarn/pnpm?

Reply:
This is removed in the next PR, but we required node at this point so yes